RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+      if (tid < nRHOther) {
+        // Rechit distance calculation
+        dist2 = (clusterPos.x - rhPos.x) * (clusterPos.x - rhPos.x) +
+                (clusterPos.y - rhPos.y) * (clusterPos.y - rhPos.y) +
+                (clusterPos.z - rhPos.z) * (clusterPos.z - rhPos.z);
+
+        d2 = dist2 / pfClusParams.showerSigma2();
+        fraction = clusterEnergy * rhENormInv * expf(-0.5 * d2);
+
+        // For single seed clusters, rechit fraction is either 1 (100%) or -1 (not included)
+        if (fraction > pfClusParams.minFracTot() && d2 < 100.)
+          fraction = 1.;
+        else
+          fraction = -1.;
+        fracView[rhFracOffset].frac() = fraction;
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if constexpr (debug) {
+        if (once_per_block(acc))
+          printf("Computing cluster position for topoId %d\n", topoId);
+      }
+
+      if (once_per_block(acc)) {
+        // Reset cluster position and energy
+        clusterPos = seedPos;
+        clusterEnergy = seedEnergy;
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      // Recalculate cluster position and energy
+      if (fraction > -0.5) {
+        alpaka::atomicAdd(acc, &clusterEnergy, rhEnergy, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.x, rhPos.x * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.y, rhPos.y * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.z, rhPos.z * rhPosNorm, alpaka::hierarchy::Blocks{});
+        alpaka::atomicAdd(acc, &clusterPos.w, rhPosNorm, alpaka::hierarchy::Blocks{});  // position_norm
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (once_per_block(acc)) {
+        // Normalize the seed postiion
+        if (clusterPos.w >= pfClusParams.minAllowedNormalization()) {
+          // Divide by position norm
+          clusterPos.x /= clusterPos.w;
+          clusterPos.y /= clusterPos.w;
+          clusterPos.z /= clusterPos.w;
+
+          if constexpr (debug)
+            printf("\tPF cluster (seed %d) energy = %f\tposition = (%f, %f, %f)\n",
+                   i,
+                   clusterEnergy,
+                   clusterPos.x,
+                   clusterPos.y,
+                   clusterPos.z);
+        } else {
+          if constexpr (debug)
+            printf("\tPF cluster (seed %d) position norm (%f) less than minimum (%f)\n",
+                   i,
+                   clusterPos.w,
+                   pfClusParams.minAllowedNormalization());
+          clusterPos.x = 0.;
+          clusterPos.y = 0.;
+          clusterPos.z = 0.;
+        }
+        float diff2 = dR2(prevClusterPos, clusterPos);
+        if constexpr (debug)
+          printf("\tPF cluster (seed %d) has diff2 = %f\n", i, diff2);
+        prevClusterPos = clusterPos;  // Save clusterPos
+
+        float diff = sqrtf(diff2);
```

Comment:
Rather that doing `diff = sqrtf(diff2)` and compare `diff` with `tol`, it's cheaper to compare `diff2` with the square of `tol`:
```c++
float tol2 = tol * tol;
notDone = (diff2 > tol2) && (iter < pfClusParams.maxIterations());
// etc.
```